db/sqlinterfaces.c

```diff
@@ -7755,16 +7757,35 @@ static int process_set_commands(struct sqlclntstate *clnt)
                     rc = ii + 1;
                 }
             } else if (strncasecmp(sqlstr, "user", 4) == 0) {
+                char *savptr = sqlstr;
                 sqlstr += 4;
                 sqlstr = cdb2_skipws(sqlstr);
-                clnt->have_user = 1;
-                strncpy(clnt->user, sqlstr, sizeof(clnt->user));
+                sqlite3Dequote(sqlstr);
+                if (strlen(sqlstr) >= sizeof(clnt->user)) {
+                    snprintf(err, sizeof(err), "set user: '%s' exceeds %d characters",
+                           sqlstr, sizeof(clnt->user) - 1);
+                    sqlstr = savptr;
+                    rc = ii + 1;
+                } else {
+                    clnt->have_user = 1;
+                    strncpy(clnt->user, sqlstr, sizeof(clnt->user) - 1);
```

Comment:
strcpy should be safe here, we have already checked the size. That makes the below line redundant too. Ditto for password.